### PR TITLE
[hcm] Add scoped RDS routing into HCM

### PR DESCRIPTION
**Pull request by stevenzzzz**
Description: add Scoped RDS routing logic into HCM. Changes include:
- in ActiveStream constructor latch a ScopedConfig impl to the activeStream if SRDS is enabled
- in the beginning of ActiveStream::decodeHeaders(headers, end_stream), get routeConfig from latched ScopedConfig impl.

This PR is the 3rd in the srds impl PR chain: [#7704, #7451, this].

Risk Level: Medium 
Testing: unit test and integration tests.
Docs Changes: N/A
Release Notes: Add scoped RDS routing support into HCM.
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by stevenzzzz** [comment]
/assign AndresGuedez
PTAL

**Comment by AndresGuedez** [comment]
Can you please merge master now that #7704 has been merged?

**Comment by stevenzzzz** [comment]
> Can you please merge master now that #7704 has been merged?

File: source/common/http/conn_manager_impl.cc [hunk 1]
```diff
@@ -593,6 +607,35 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
...
+      connection_manager_.config_.scopedRouteConfigProvider() != nullptr) {
+    ASSERT(snapped_route_config_ == nullptr,
+           "Route config already latched to the active stream when scoped RDS is enabled.");
+    if (snapped_scoped_routes_config_ == nullptr) {
+      ENVOY_STREAM_LOG(trace, "snapped scoped routes config is null when SRDS is enabled.", *this);
+      // Stop decoding now.
+      maybeEndDecode(true);
+      sendLocalReply(
+          Grpc::Common::hasGrpcContentType(*request_headers_), Http::Code::InternalServerError,
+          "scoped routes config not set when SRDS is enabled", nullptr, is_head_request_,
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/http/conn_manager_impl.cc]: sounds good, will address it here. 
(This is a test to see if this PR can be opopened)

File: test/integration/scoped_rds_integration_test.cc [hunk 2]
```diff
@@ -104,29 +152,52 @@ class ScopedRdsIntegrationTest : public HttpIntegrationTest,
...
   }
 
-  void createRdsStream() { createStream(&rds_upstream_info_, getRdsFakeUpstream()); }
+  void createRdsStream(const std::string& resource_name) {
+    createStream(&rds_upstream_info_, getRdsFakeUpstream(), resource_name);
+  }
 
   void createScopedRdsStream() {
-    createStream(&scoped_rds_upstream_info_, getScopedRdsFakeUpstream());
+    createStream(&scoped_rds_upstream_info_, getScopedRdsFakeUpstream(), "foo-scoped-routes");
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/integration/scoped_rds_integration_test.cc]: done

File: source/common/http/conn_manager_impl.cc [hunk 3]
```diff
@@ -593,6 +607,35 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(HeaderMapPtr&& headers,
   ScopeTrackerScopeState scope(this,
                                connection_manager_.read_callbacks_->connection().dispatcher());
   request_headers_ = std::move(headers);
+  // For Admin thread, we don't use routeConfigProvider or SRDS route provider.
+  if (dynamic_cast<Server::Admin*>(&connection_manager_.config_) == nullptr &&
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/http/conn_manager_impl.cc]: this is not route-selection yet. It's just attaching RouteConfiguration to the activeStream. 
Previously for RouteConfigProvider, the snapping of RouteConfiguration happens at ActiveStream creation time.
For SRDS, we have to wait till decoderheaders gets called tho.

Snapping the RouteConfiguration has to happen before the actual routeSelection, as there are data from Route used in other places (for now it's used in header cleanup, around "stream_info_.setDownstreamRemoteAddress(ConnectionManagerUtility::mutateRequestHeaders").

File: test/integration/scoped_rds_integration_test.cc [hunk 4]
```diff
@@ -150,49 +221,146 @@ INSTANTIATE_TEST_SUITE_P(IpVersionsAndGrpcTypes, ScopedRdsIntegrationTest,
...
+  // Test "foo-route" and 'bar-route' both gets routed to cluster_0.
+  test_server_->waitForCounterGe("http.config_test.rds.foo_route1.update_success", 1);
+  for (const std::string& scope_key : std::vector<std::string>{"foo-route", "bar-route"}) {
+    sendRequestAndVerifyResponse(
+        Http::TestHeaderMapImpl{{":method", "GET"},
+                                {":path", "/meh"},
+                                {":authority", "host"},
+                                {":scheme", "http"},
+                                {"Addr", fmt::format("x-foo-key={}", scope_key)}},
+        456, Http::TestHeaderMapImpl{{":status", "200"}, {"service", scope_key}}, 123,
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/integration/scoped_rds_integration_test.cc]: The upstream sets it around "response = sendRequestAndWaitForResponse(request_headers, request_size, response_headers,"

File: source/common/http/conn_manager_impl.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply source/common/http/conn_manager_impl.cc]: done

File: test/mocks/router/mocks.h [hunk 5]
```diff
@@ -379,16 +382,29 @@ class MockConfig : public Config {
...
+  MOCK_METHOD0(onConfigUpdate, void());
+
+  std::shared_ptr<NiceMock<MockConfig>> route_config_{new NiceMock<MockConfig>()};
+};
+
 class MockRouteConfigProviderManager : public RouteConfigProviderManager {
 public:
-  MockRouteConfigProviderManager();
-  ~MockRouteConfigProviderManager() override;
+  MockRouteConfigProviderManager() = default;
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/mocks/router/mocks.h]: reverted.
I don't have data but leaving it inline saves a call-to, which might be even faster considering these classes are almost empty.

File: test/common/http/conn_manager_impl_test.cc [hunk 6]
```diff
@@ -4359,5 +4371,169 @@ TEST_F(HttpConnectionManagerImplTest, TestSessionTrace) {
   }
 }
 
+TEST_F(HttpConnectionManagerImplTest, TestSRDSRouteNotFound) {
+  setup(false, "");
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/common/http/conn_manager_impl_test.cc]: done.

File: test/common/http/conn_manager_impl_test.cc [hunk 7]
```diff
@@ -4359,5 +4371,169 @@ TEST_F(HttpConnectionManagerImplTest, TestSessionTrace) {
...
+  use_srds_ = true;
+
+  EXPECT_CALL(*static_cast<const Router::MockScopedConfig*>(
+                  scopedRouteConfigProvider()->config<Router::ScopedConfig>().get()),
+              getRouteConfig(_))
+      .Times(1)
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/common/http/conn_manager_impl_test.cc]: done here and everywhere in this file.

File: test/common/http/conn_manager_impl_test.cc [hunk 8]
```diff
@@ -4359,5 +4371,169 @@ TEST_F(HttpConnectionManagerImplTest, TestSessionTrace) {
...
+    StreamDecoder* decoder = &conn_manager_->newStream(response_encoder_);
+    HeaderMapPtr headers{
+        new TestHeaderMapImpl{{":authority", "host"}, {":method", "GET"}, {":path", "/foo"}}};
+    decoder->decodeHeaders(std::move(headers), true);
+    data.drain(4);
+  }));
+
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, false))
+      .WillOnce(Invoke([](const HeaderMap& headers, bool) -> void {
+        EXPECT_EQ("404", headers.Status()->value().getStringView());
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/common/http/conn_manager_impl_test.cc]: pardon me make a defense here. I actually think this makes the test easier to read: putting it in a local variables and verify it later make it harder to track where the value is assigned, and expect_call could be put somewhere far from the actual call-site (onData).
I noticed there are these two patterns in this files already, if I have to choose, I side with the "putting check closer to its call-site" way.
If you don't feel strongly, I will keep it this way. :)

File: test/common/http/conn_manager_impl_test.cc [hunk 9]
```diff
@@ -4359,5 +4371,169 @@ TEST_F(HttpConnectionManagerImplTest, TestSessionTrace) {
...
+
+  scoped_route_config_provider2_ =
+      std::make_shared<NiceMock<Router::MockScopedRouteConfigProvider>>();
+  // Can't have RDS and SRDS provider in the same time.
+  EXPECT_DEBUG_DEATH(conn_manager_->onData(fake_input, false),
+                     "Either routeConfigProvider or scopedRouteConfigProvider should be set in "
+                     "ConnectionManagerImpl.");
+
+  scoped_route_config_provider2_.reset();
+  // Only scoped route config provider valid.
```
[the remark this answers is not in the document]
Comment (stevenzzzz) [reply test/common/http/conn_manager_impl_test.cc]: yeah, thanks for the catch!

**Comment by stevenzzzz** [comment]
Added delta api support in integration test, PTAL.

@htuch  @AndresGuedez

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 10]
```diff
@@ -50,6 +50,20 @@ request. The router filter supports the following features:
 * :ref:`Hash policy <envoy_api_field_route.RouteAction.hash_policy>` based routing.
 * :ref:`Absolute urls <envoy_api_field_config.filter.network.http_connection_manager.v2.HttpConnectionManager.http_protocol_options>` are supported for non-tls forward proxies.
 
+
+Routing Scope
+--------------
+
+Scoped routing enables Envoy to put constraints on search space of domains and route rules.
+A :ref:`Routeing Scope<envoy_api_msg_ScopedRouteConfiguration>` associates a key with a :ref:`route table <arch_overview_http_routing_route_table>`.
```
Comment (AndresGuedez) [inline docs/root/intro/arch_overview/http/http_routing.rst]: nit: s/Routeing/Routing/

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 11]
```diff
@@ -50,6 +50,34 @@ request. The router filter supports the following features:
...
+For each request, a scope key is computed dynamically by the HTTP connection manager to pick the :ref:`route table<envoy_api_msg_RouteConfiguration>`.
+
+The Scoped RDS (SRDS) API contains a set of :ref:`Scopes <envoy_api_msg_ScopedRouteConfiguration>` resources, each defining independent routing configuration,
+along with a :ref:`ScopeKeyBuilder <envoy_api_msg_config.filter.network.http_connection_manager.v2.ScopedRoutes.ScopeKeyBuilder>`
+defining the key construction algorithm used by Envoy to look up the scope corresponding to each request.
+
+For example, for the following scoped route configuration, Envoy  will look into the "addr" header value, split the header value by ";" first, and use the first value for key 'x-foo-key' as the scope key.
```
Comment (htuch) [inline docs/root/intro/arch_overview/http/http_routing.rst]: Tiny nit: there is a weird double space here.

File: docs/root/intro/arch_overview/http/http_routing.rst [hunk 12]
```diff
@@ -50,6 +50,34 @@ request. The router filter supports the following features:
...
+If the "addr" header value is "foo=1;x-foo-key=127.0.0.1;x-bar-key=1.1.1.1", then "127.0.0.1" will be computed as the scope key to look up for corresponding route configuration.
+
+.. code-block:: yaml
+
+  name: 'scope_by_addr'
```
Comment (htuch) [inline docs/root/intro/arch_overview/http/http_routing.rst]: Nit: no quotes needed for YAML string value.

File: docs/root/intro/arch_overview/http/http_routing.rst (hunk 10 above)
Comment (stevenzzzz) [reply docs/root/intro/arch_overview/http/http_routing.rst]: oops

File: docs/root/intro/arch_overview/http/http_routing.rst (hunk 11 above)
Comment (stevenzzzz) [reply docs/root/intro/arch_overview/http/http_routing.rst]: done

File: docs/root/intro/arch_overview/http/http_routing.rst (hunk 12 above)
Comment (stevenzzzz) [reply docs/root/intro/arch_overview/http/http_routing.rst]: done

File: docs/root/intro/version_history.rst [hunk 13]
```diff
@@ -53,6 +53,7 @@ Version history
 * upstream: use p2c to select hosts for least-requests load balancers if all host weights are the same, even in cases where weights are not equal to 1.
 * upstream: added :ref:`an option <envoy_api_field_Cluster.CommonLbConfig.close_connections_on_host_set_change>` that allows draining HTTP, TCP connection pools on cluster membership change.
 * zookeeper: parse responses and emit latency stats.
+* router: :ref:`Scoped routing <arch_overview_http_routing_route_scope>` is supported.
```
Comment (htuch) [inline docs/root/intro/version_history.rst]: Need to be alpha sorted, otherwise LGTM.

Comment (stevenzzzz) [reply docs/root/intro/version_history.rst]: done.

**Comment by htuch** [comment]
`listener_impl_test` failure is almost definitely unrelated, so will force merge.

**Comment by fredlas** [comment]
@htuch @stevenzzzz it looks like this PR broke test/integration:ads_integration_test for https://github.com/envoyproxy/envoy/pull/7293, which has had it passing since mid-June, when it was first ready to merge. (This test had one or two merge conflicts along the way, but nothing major).

Specifically, the failures begin halfway through the basic test flow (ads_integration.cc:235) in the very first basic test. An EDS request is made for cluster_1,cluster_0 when cluster_2,cluster_1 were what was expected. There are then a bunch more failures, but it looks like they're all just due to the type_url getting shifted out of whack. It looks like the desired {2,1} request does actually happen right after; it's just that the {1,0} is inserted before it.

So, the failure itself seems somewhat straightforward, but since it was caused by something in this rather large change, which I'm not at all familiar with, I'm going to need some help. (Which might be as simple as getting confirmation that it's valid to just add an expectation for the new message. Or maybe even relax the message-by-message expectations, in favor of a more end-to-end test of the config behavior?)

File: source/common/http/conn_manager_impl.cc [hunk 14]
```diff
@@ -12,6 +12,7 @@
 #include "envoy/event/dispatcher.h"
 #include "envoy/network/drain_decision.h"
 #include "envoy/router/router.h"
+#include "envoy/server/admin.h"
```
Comment (mattklein123) [inline source/common/http/conn_manager_impl.cc]: drive by follow up comment: I'm not crazy about including admin here and then special casing it with dynamic casts. Is it not possible for this logical to be external to the construction of the HCM and then specified correctly either by admin or by the normal filter?

Comment (stevenzzzz) [reply source/common/http/conn_manager_impl.cc]: we've discussed it offline, I will make a cleanup/refactor PR to move the dynamic_cast outside of conn_manager_impl.

**Comment by stevenzzzz** [comment]
> @htuch @stevenzzzz it looks like this PR broke test/integration:ads_integration_test for #7293, which has had it passing since mid-June, when it was first ready to merge. (This test had one or two merge conflicts along the way, but nothing major).
> 
> Specifically, the failures begin halfway through the basic test flow (ads_integration.cc:235) in the very first basic test. An EDS request is made for cluster_1,cluster_0 when cluster_2,cluster_1 were what was expected. There are then a bunch more failures, but it looks like they're all just due to the type_url getting shifted out of whack. It looks like the desired {2,1} request does actually happen right after; it's just that the {1,0} is inserted before it.
> 
> So, the failure itself seems somewhat straightforward, but since it was caused by something in this rather large change, which I'm not at all familiar with, I'm going to need some help. (Which might be as simple as getting confirmation that it's valid to just add an expectation for the new message. Or maybe even relax the message-by-message expectations, in favor of a more end-to-end test of the config behavior?)

This is caused by the change in PR7293 source/common/upstream/cds_api_impl.cc where the CleanUp object would immediately unpause the Cluster EDS, which caused the list of clusters to be sent still contain the ( 0) which was deleted.
```
  if (cm_.adsMux()) {                                                                                                                                                                                                                                                         
    cm_.adsMux()->pause(Config::TypeUrl::get().ClusterLoadAssignment);                                                                                                                                                                                                        
    Cleanup eds_resume(                                                                                                                                                                                                                                                       
        [this] { cm_.adsMux()->resume(Config::TypeUrl::get().ClusterLoadAssignment); });                                                                                                                                                                                      
  }   
``` 
A trick is to use a unique_ptr<CleanUp> outside of the if-statement, and create a CleanUp in the statement.

**Pull request merged**